### PR TITLE
[move-ide] Added a skeleton of the Prettier Move plugin

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This PR adds a skeleton of Move plugin for [Prettier](https://github.com/prettier/prettier). It only supports a few constructs at the moment but it has been verified that it works both on the command line and in VSCode, and it's platform independent due to it using a WASM-ified version of tree-sitter's Move parser.

## Test Plan 

Manually tested that it works on the command line and with VSCode. Added a testing framework to test individual features and added tests for existing features.

### CODE REVIEW

**Review by tnowacki — COMMENTED** [review]
What's the story for the wasm bit? will we pull that in? How do we generate it as a build artifact and keep it checked in?

**Review by tnowacki — COMMENTED** [review]
Overall very exciting! and I don't really have many comments because ya know... javascript/typescript 🤷 
And we can chat about how to proceed with the spec or w/e

**Comment by awelc** [comment]
> What's the story for the wasm bit? will we pull that in? How do we generate it as a build artifact and keep it checked in?

It's [described](https://github.com/MystenLabs/sui/tree/aw/prettier-plugin/external-crates/move/crates/move-analyzer/prettier-plugin#prettier-move-plugin) in the README - there is a [script](https://github.com/MystenLabs/sui/blob/aw/prettier-plugin/external-crates/move/crates/move-analyzer/prettier-plugin/scripts/treesitter-wasm-gen.sh) in the `scripts` subdir that pulls @tzakian's treesitter Move parser repo into a temp directory and builds the WASM bit. As noted in the README, one has to be careful when doing so as blindly updating the WASM bit may break the existing formatter (as the formatter relies on a certain parse tree structure, and if the tree changes in a non-compatible fashion, it can break things). The idea is that updates would happen rarely and by-hand rather than automatically.

File: external-crates/move/crates/move-analyzer/prettier-plugin/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+export const parsers = {
+    'move-parse': {
+        parse: (text: string): Promise<SyntaxNode> => {
+            return (async (): Promise<SyntaxNode> => {
+                await Parser.init();
+                const parser = new Parser();
+                const Lang = await Parser.Language.load(path.join(__dirname,
+                                                                  '..',
+                                                                  'tree-sitter-move.wasm'));
```
Comment (rvantonder) [inline external-crates/move/crates/move-analyzer/prettier-plugin/src/index.ts:27]: nit: strange formatting, maybe we need a different prettier config for these TS files? 😄

File: external-crates/move/crates/move-analyzer/prettier-plugin/src/test_run.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+if (process.argv.length !== 3) {
+    usage();
+    process.exit(1);
+}
+
+const filepath = process.argv[2]
+
+if (filepath != null) {
+
```
Comment (rvantonder) [inline external-crates/move/crates/move-analyzer/prettier-plugin/src/test_run.ts:23]:
```suggestion
```

File: external-crates/move/crates/move-analyzer/prettier-plugin/tests/run_spec.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+const assert = require('assert');
+const linediff = require('line-diff');
+const fs = require('fs');
+const path = require('path');
+const plugin = require('../');
+const prettier = require('prettier')
```
Comment (rvantonder) [inline external-crates/move/crates/move-analyzer/prettier-plugin/tests/run_spec.js:9]: FYI [jest](https://jestjs.io/docs/expect#tomatchsnapshotpropertymatchers-hint) is a convenient snapshot testing library that might be nice to have later (also does the diffing, etc.)

Not suggesting we add this now, just putting it on your radar :-)

**Review by rvantonder — APPROVED** [review]
Looks good, let's ship this!

File: external-crates/move/crates/move-analyzer/prettier-plugin/src/index.ts (hunk 1 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/prettier-plugin/src/index.ts]: Admittedly, it was formatted by hand... :-) Changed it to something hopefully more palatable.

**Comment by dariorussi** [comment]
nice thanks!!!

**Pull request merged**